.github/workflows/linux.yml

```diff
@@ -1,51 +1,46 @@
 name: Linux builds
 on: push
-jobs:
+env:
+  MAX_WARNINGS_GCC_6_Debug:     378
+  MAX_WARNINGS_GCC_6_Release:   661
+  MAX_WARNINGS_GCC_7_Debug:     380
+  MAX_WARNINGS_GCC_7_Release:   329
+  MAX_WARNINGS_GCC_8_Debug:     380
+  MAX_WARNINGS_GCC_8_Release:   330
+  MAX_WARNINGS_GCC_9_Debug:     403
+  MAX_WARNINGS_GCC_9_Release:   358
+  MAX_WARNINGS_Clang_7_Debug:   193
+  MAX_WARNINGS_Clang_7_Release: 193
+  MAX_WARNINGS_Clang_8_Debug:   193
+  MAX_WARNINGS_Clang_8_Release: 193
 
-  build_ubuntu_clang_8:
-    name: Clang 8 (ubuntu-18.04)
-    runs-on: ubuntu-18.04
-    steps:
-      - uses: actions/checkout@v1
-      - name: Log environment
-        run:  ./scripts/log-env.sh
-      - run:  sudo apt-get update
-      - name: Install dependencies
-        run:  sudo apt-get install -y $(./scripts/list-build-dependencies.sh -p apt --compiler clang --compiler-version 8)
-      - name: Build
-        run:  ./scripts/build.sh --compiler clang --compiler-version 8 --lto
-      - name: Summarize warnings
-        run:  ./scripts/count-warnings.py build.log
+jobs:
 
-  build_ubuntu_matrix_gcc:
-    name: GCC
-    runs-on: ${{ matrix.os }}
+  build_matrix:
+    name: ${{ matrix.compiler }}-${{ matrix.compiler_version }}
+    runs-on: ubuntu-latest
```

Comment:
For Windows and macOS we were forced to use `*-latest`, but for Ubuntu, we are not. I am ok with replacing `ubuntu-18.04` with `ubuntu-latest`, but we need one job, which explicitly uses the oldest GCC still available and supported on GitHub CI.

This is quite important, as we'll need to use that build configuration to prepare "release builds" to share with testers. I know this seems counter-intuitive (why use 16.04 when 18.04 is newer?), but when sharing a builds with end users/testers, going with older distro gives wider range of compatibility (with Ubuntu + derivatives, Debian + derivatives and even other, totally unrelated distributions).

This is an error often repeated e.g. by game developers or companies trying to package proprietary software for Linux - they try to use the newer distro (because that is beneficial for development), and when user problems start seeping in, they blame "fragmentation".

One more reason to keep Ubuntu 16.04 as long as possible is SteamOS. I am working on dosbox-staging mainly with intention of bundling it with Boxtron for SteamOS builds, and atm, non-beta Steam runtime environment offers only GCC 5.4 (coincidentally, the same version as used in Ubuntu 16.04).

Reply:
Excellent point and something that slipped my mind when I dropped 16.04. 

Even if we can use an old compiler on the latest Linux, the libc bindings will be against the current version, so I think we need to use the oldest-Linux + oldest-compiler for releases.  Or we can trying linking more of these core libraries statically.

Will re-add 16.04 with the default compiler, and will knock out the corresponding compiler version in Ubuntu latest.

Reply:
`libc` won't be a problem - very rarely, an old, unsupported libc can cause a problem when compiling new code, but I doubt it we will hit such situation (I actually did hit it when compiling Vulkan renderer for GZDoom for SteamOS, but it was easy to fix it upstream). `libc++` is a bigger risk, but still low (I saw some developers trying to statically link either of those to get e.g. C++17 support on old distro, but this is very bad idea). The biggest risk is one of dependency libraries missing ABI version used for compilation on newer system (I experienced it only only on less-popular distros such as Solus or NixOS - mainstream ones have this base covered usually, at least for popular, high quality libraries). 

In practice: we just need to test it, and I am quite sure using package compiled on Ubuntu 18.04 will fail (that's what DOSBox ECE is doing and that package is unusable for me).

Reply:
Sounds good, and glad you have modern experience with distributing unix binaries under modern environments.  Last I did stuff like this was early 2000's, and libc was in heavy flux.  

For my audio patch on vogons, I have a separate tool I've wrote to build and all dependencies from source, and link dosbox fully statically (libz, libpng, sdl, etc), even on windows. The binaries are 30+MB, and everyone can run them, bur definitely excessive!

Same here regarding ECE.. the binary is unusable for me (32bit I believe too?).

Reply:
Hmm, this gives me an idea: Git (project) uses a [`contrib/`](https://github.com/git/git/tree/master/contrib) directory for all scripts and programs, that are either not strictly part of Git or are "maturing" before being moved out of that dir, and being distributed with Git. Perhaps we should adopt this idea? I would be less strict about reviews for adjunct scripts to that directory, and it would be good place to organize other files, that we should include with DOSBox (e.g. Fedora provides README files translated into several languages by the community  - those should be included in repo).

Reply:
Yes; `contrib/` directories are normally packed with goodies that help various users, but may not be core to the software or it's core-development. I've been a heavy user of contrib/ scripts in gcc's sources (they have a script that downloads and builds all dependencies, like gmp, mpfr, etc.. ); same with the linux kernel - where I used scripts to generate firmware blobs for ARM, and have used them in the past when I built Firefox and Mozilla from sources (a crazy prospect given the hardware power back in those days!).

Reply:
Fixed in d28cc6bb